### PR TITLE
Windows: Additional write permissions check

### LINKED ISSUE #3910: `You have no permission to write to the selected folder` checkPathValidityForNewFolder failes for SMB-folders
I am using the Windows Desktop Client to sync a folder located on a SMB share to a self hosted nextcloud instance. This worked for Versions up to 3.1.3. Starting with 3.2.0-rc1 (tested up to 3.4.4) a error ("`You have no permission to write to the selected folder!`") appears when adding a folder synchronization pair. My account has full write permission for the share/folder, creating files using Windows Explorer, Folder Wizard (new folder) or Settings-GUI (save debug logs) works fine. Folders located on local disks work as expected.

Ref: https://help.nextcloud.com/t/you-have-no-permission-to-write-to-the-selected-folder/125486

### Expected behaviour
Add writable SMB-folder to sync pair

### Actual behaviour
Error `You have no permission to write to the selected folder!` is displayed while writing should work fine. Next-button is disabled.

### Steps to reproduce
1. Map a NAS drive using SMB ("Windows file share") to a local drive letter (Tested targets: Samba 4.9.x, Windows Server 2012R2, Windows 10)
2. Install Windows Client 3.2.0 - 3.4.4
3. Connect to a NC Server
4. When asked for a local directory browse to the mapped drive, create an empty folder using the NC GUI (aka: prove it is writable) and try to select it

### Client configuration
Client version:
* 3.2.0-rc1 (x64)
* 3.3.5stable-Win64 (build 20210930); 0d9a6987ebb0040b34451e51f54121dbd339ddf8; Sep 30 2021, 13:30:50; Qt 5.12.10, OpenSSL 1.1.1g  21 Apr 2020
* (Answers: 3.4.1-3.4.4)

Operating system:
* Windows 10 Pro
* Windows 11
* (Forum: Windows 8.1)

OS language:
* German
* English

Installation path of client:
Default (C:\Program Files\Nextcloud)

### Logs
1. Client logfile (Win10 German 3.3.5): 
```
2021-10-19 21:34:24:941 [ info nextcloud.gui.folder.manager ]:	Folders to sync: 0
2021-10-19 21:34:24:941 [ info nextcloud.gui.folder.manager ]:	Number of folders that don't use push notifications: 0
2021-10-19 21:34:37:263 [ debug nextcloud.gui.folder.manager ]	[ OCC::FolderMan::checkPathValidityForNewFolder ]:	"d" "Sie haben keine Schreibberechtigung für den ausgewählten Ordner!"
2021-10-19 21:34:37:263 [ debug nextcloud.gui.folder.manager ]	[ OCC::FolderMan::checkPathValidityForNewFolder ]:	"d" "Sie haben keine Schreibberechtigung für den ausgewählten Ordner!"
2021-10-19 21:34:37:264 [ debug nextcloud.gui.folder.manager ]	[ OCC::FolderMan::checkPathValidityForNewFolder ]:	"d" "Sie haben keine Schreibberechtigung für den ausgewählten Ordner!"
2021-10-19 21:34:37:440 [ debug nextcloud.gui.folder.manager ]	[ OCC::FolderMan::checkPathValidityForNewFolder ]:	"d:" "Sie haben keine Schreibberechtigung für den ausgewählten Ordner!"
2021-10-19 21:34:37:666 [ debug nextcloud.gui.folder.manager ]	[ OCC::FolderMan::checkPathValidityForNewFolder ]:	"d:/" "Sie haben keine Schreibberechtigung für den ausgewählten Ordner!"
2021-10-19 21:34:38:317 [ debug nextcloud.gui.folder.manager ]	[ OCC::FolderMan::checkPathValidityForNewFolder ]:	"d:/o" "Sie haben keine Schreibberechtigung für den ausgewählten Ordner!"
2021-10-19 21:34:38:423 [ debug nextcloud.gui.folder.manager ]	[ OCC::FolderMan::checkPathValidityForNewFolder ]:	"d:/ow" "Sie haben keine Schreibberechtigung für den ausgewählten Ordner!"
2021-10-19 21:34:38:643 [ debug nextcloud.gui.folder.manager ]	[ OCC::FolderMan::checkPathValidityForNewFolder ]:	"d:/own" "Sie haben keine Schreibberechtigung für den ausgewählten Ordner!"
2021-10-19 21:34:52:133 [ debug nextcloud.gui.folder.manager ]	[ OCC::FolderMan::checkPathValidityForNewFolder ]:	"D:/Owncloud-Test/test2" "Sie haben keine Schreibberechtigung für den ausgewählten Ordner!"
2021-10-19 21:34:52:208 [ debug nextcloud.gui.folder.manager ]	[ OCC::FolderMan::checkPathValidityForNewFolder ]:	"D:/Owncloud-Test/test2" "Sie haben keine Schreibberechtigung für den ausgewählten Ordner!"
2021-10-19 21:34:52:279 [ debug nextcloud.gui.folder.manager ]	[ OCC::FolderMan::checkPathValidityForNewFolder ]:	"D:/Owncloud-Test/test2" "Sie haben keine Schreibberechtigung für den ausgewählten Ordner!"
2021-10-19 21:34:54:942 [ info nextcloud.gui.folder.manager ]:	Etag poll timer timeout
2021-10-19 21:34:54:942 [ info nextcloud.gui.folder.manager ]:	Folders to sync: 0
2021-10-19 21:34:54:942 [ info nextcloud.gui.folder.manager ]:	Number of folders that don't use push notifications: 0
2021-10-19 21:34:56:244 [ info nextcloud.gui.account.settings ]:	Folder wizard cancelled
```

### PR DESCRIPTION
**Pull request by Alkl58**
The Problem: The [isWritable()](https://doc.qt.io/qt-6/qfileinfo.html#isWritable) function checks only(?) for write [permissions](https://doc.qt.io/qt-6/qfiledevice.html#Permission-enum) for the user, owner, group or anyone(?). 

**Its technically correct** - when opening the folder permissions of an SMB share in Explorer, it shows that the user doesn't have those permissions. The actual permission for being able to read/write/delete is not covered by the "base" permissions of Explorer, it's covered by the "special permissions". 

It looks similar to this (just imageine it being an SMB share):
 

**For more info check my comment:** https://github.com/nextcloud/desktop/issues/3910#issuecomment-1460924930

This PR adds an additional actual write test for Windows hosts.

This PR fixes #3910 

_First time using C++, comments/improvements welcome!_

### CODE REVIEW

**Comment by Alkl58** [comment]
Also closes/fixes: #3201

**Comment by Alkl58** [comment]
Test on SMB share:
grafik

Test on normal folder:
grafik

Test on protected system folder:
grafik

Test on write protected folder:
grafik

**Comment by Alkl58** [comment]
Will this be merged at some point? Or should I close it?

**Review by claucambra — COMMENTED** [review]
It looks good at first glance, but need to find time to test

File: src/gui/folderman.cpp [hunk 1]
```diff
@@ -1664,9 +1664,27 @@ static QString checkPathValidityRecursive(const QString &path)
         return FolderMan::tr("The selected path is not a folder!");
     }
 
+    #ifdef Q_OS_WIN
+    if (!selFile.isWritable()) {
+        // isWritable() doesn't cover all NTFS permissions
+        // Try write
+        auto testPath = path.toStdString() + "\\nextcloud-write-test-file.txt";
+        FILE *fp = fopen(testPath.c_str(), "w");
```
Comment (claucambra) [inline src/gui/folderman.cpp:1672]:
```suggestion
        const auto fp = fopen(testPath.c_str(), "w");
```

File: src/gui/folderman.cpp [hunk 2]
```diff
@@ -1664,9 +1664,27 @@ static QString checkPathValidityRecursive(const QString &path)
         return FolderMan::tr("The selected path is not a folder!");
     }
 
+    #ifdef Q_OS_WIN
+    if (!selFile.isWritable()) {
+        // isWritable() doesn't cover all NTFS permissions
+        // Try write
+        auto testPath = path.toStdString() + "\\nextcloud-write-test-file.txt";
```
Comment (claucambra) [inline src/gui/folderman.cpp:1671]:
```suggestion
        const auto testPath = path.toStdString() + "\\nextcloud-write-test-file.txt";
```

File: src/gui/folderman.cpp [hunk 3]
```diff
@@ -1664,9 +1664,27 @@ static QString checkPathValidityRecursive(const QString &path)
...
+        FILE *fp = fopen(testPath.c_str(), "w");
+        if (!fp)
+            return FolderMan::tr("You have no permission to write to the selected folder!");
```
Comment (claucambra) [inline src/gui/folderman.cpp:1674]:
```suggestion
        if (!fp) {
            return FolderMan::tr("You have no permission to write to the selected folder!");
        }
```

File: src/gui/folderman.cpp [hunk 4]
```diff
@@ -1664,9 +1664,27 @@ static QString checkPathValidityRecursive(const QString &path)
...
+        fclose(fp);
+
+        // Try delete
+        auto rc = remove(testPath.c_str());
```
Comment (claucambra) [inline src/gui/folderman.cpp:1678]:
```suggestion
        const auto rc = remove(testPath.c_str());
```

File: src/gui/folderman.cpp [hunk 5]
```diff
@@ -1664,9 +1664,27 @@ static QString checkPathValidityRecursive(const QString &path)
...
+        if (rc)
+            return FolderMan::tr("You have no permission to write to the selected folder!");
```
Comment (claucambra) [inline src/gui/folderman.cpp:1680]:
```suggestion
        if (rc) {
            return FolderMan::tr("You have no permission to write to the selected folder!");
        }
```

**Review by claucambra — CHANGES_REQUESTED** [review]
A couple of small comments

**Review by claucambra — CHANGES_REQUESTED** [review]
Hi @Alkl58 are you able to make the requested changes? If not we are happy to take over the PR and include your changes with our own fixes :)

**Comment by Alkl58** [comment]
> Hi @Alkl58 are you able to make the requested changes? If not we are happy to take over the PR and include your changes with our own fixes :)

Sorry for the delay. Should be all good now :)

File: src/gui/folderman.cpp [hunk 6]
```diff
@@ -1701,9 +1701,29 @@ static QString checkPathValidityRecursive(const QString &path)
         return FolderMan::tr("The selected path is not a folder!");
     }
 
+    #ifdef Q_OS_WIN
+    if (!selFile.isWritable()) {
+        // isWritable() doesn't cover all NTFS permissions
+        // Try write
+        const auto testPath = selFile.dir().filePath("nextcloud-write-test-file.txt");
+        const auto fp = fopen(testPath.c_str(), "w");
```
Comment (allexzander) [inline src/gui/folderman.cpp:1709]: const auto fp = fopen(testPath.toStdWString().c_str(), "w");

File: src/gui/folderman.cpp [hunk 7]
```diff
@@ -1701,9 +1701,29 @@ static QString checkPathValidityRecursive(const QString &path)
...
+        if (!fp) {
+            return FolderMan::tr("You have no permission to write to the selected folder!");
+        }
+        fclose(fp);
+
+        // Try delete
+        const auto rc = remove(testPath.c_str());
```
Comment (allexzander) [inline src/gui/folderman.cpp:1716]: const auto rc = remove(testPath.toStdWString().c_str());

**Review by allexzander — CHANGES_REQUESTED** [review]

**Comment by allexzander** [comment]
> Just a FYI: Apparently I completely broke my local dev environment after trying to install all qt dependencies. I'll reinstall my system / use a VM, will take some time so I can properly debug / fix things.

sure, no worries, I've noticed you're using `fopen` function from C, that only takes `const char*` as filename. We need to use `QFile` instead. If you are ok, I can take over the PR and incorporate changes.

**Comment by Alkl58** [comment]
@allexzander sure, thats probably faster, sorry for the hassle

**Comment by allexzander** [comment]
Moved to https://github.com/nextcloud/desktop/pull/6209

**Pull request closed without merging**